library/core/src/num/mod.rs

```diff
@@ -1033,19 +1034,41 @@ impl_helper_for! { i8 i16 i32 i64 i128 isize u8 u16 u32 u64 u128 usize }
 #[doc(hidden)]
 #[inline(always)]
 #[unstable(issue = "none", feature = "std_internals")]
-pub fn can_not_overflow<T>(radix: u32, is_signed_ty: bool, digits: &[u8]) -> bool {
+pub const fn can_not_overflow<T>(radix: u32, is_signed_ty: bool, digits: &[u8]) -> bool {
     radix <= 16 && digits.len() <= mem::size_of::<T>() * 2 - is_signed_ty as usize
 }
 
-fn from_str_radix<T: FromStrRadixHelper>(src: &str, radix: u32) -> Result<T, ParseIntError> {
+const fn from_str_radix<
+    T: ~const FromStrRadixHelper + ~const Add + ~const Sub + ~const Mul + ~const PartialOrd,
+>(
+    src: &str,
+    radix: u32,
+) -> Result<T, ParseIntError> {
     use self::IntErrorKind::*;
     use self::ParseIntError as PIE;
 
-    assert!(
-        (2..=36).contains(&radix),
-        "from_str_radix_int: must lie in the range `[2, 36]` - found {}",
-        radix
-    );
+    if 2 > radix || radix > 36 {
+        // FIXME DIY const string formatting. Doing it the normal way is blocked on something
+        // like https://github.com/rust-lang/rust/issues/86364.
+        const MESSAGE: &[u8; 60] = b"from_str_radix_int: must lie in the range `[2, 36]` - found ";
+        let mut buf = [mem::MaybeUninit::<u8>::uninit(); MESSAGE.len() + 10];
+        panic!("{}", {
+            let buf_end = buf.as_mut_ptr_range().end;
+            let radix_len = crate::fmt::fmt_u64(
+                radix.into(),
+                // SAFETY: OK to refer to the last 20 bytes of a 70 byte buffer.
+                unsafe { &mut *(buf_end.sub(20).cast()) },
+            )
+            .len();
+            let result = &mut buf[(10 - radix_len)..];
+            // SAFETY: (10-radix_len) above is in 1..=10 so 60 <= result.len() <= 69
+            unsafe {
+                MESSAGE.as_ptr().copy_to_nonoverlapping(result.as_mut_ptr().cast(), MESSAGE.len());
+                // SAFETY: result is entirely filled by MESSAGE and the radix, both ASCII.
+                crate::str::from_utf8_unchecked(mem::MaybeUninit::slice_assume_init_ref(result))
+            }
+        });
+    }
```

Comment:
FYI, there is also the way of using const_eval_select: https://github.com/rust-lang/rust/blob/f80e454450c891c075617d23532c6eb317f4471d/library/core/src/slice/index.rs#L34-L57